docs/writing-rules/testing-rules.md

```diff
@@ -31,6 +31,27 @@ Other than annotations there are three things to remember when creating tests:
 The `.test.yaml` file extension can also be used for test files. This is necessary when testing YAML language rules.
 :::
 
+## Testing autofix
+
+Semgrep's testing mechanism also provides a way to test the behaviour of any fixes defined in the rules.
+To define a test for autofix behaviour: 
+
+1. Create a new code file with the `.fixed` suffix before the annotation. 
+   For example, the autofix test file for a rule with test code in `path/to/rule.py` will be named `path/to/rule.fixed.py`.
+2. This file should contain the expected result of applying autofixes to the test code file.
+3. Any future runs of the command `semgrep --test` will apply the autofix to the original testcode (`path/to/rule.py`) and verify whether this matches the expected outcome defined in the autofix test file (`path/to/rule.fixed.py`).
+   If there is a mismatch, the line diffs will be printed.
```

Comment:
Usually we prefer to have "will" (future tense) be present tense, but I think it's okay in this case? What do you think @adamkvitek ?

Reply:
@s-santillan I would avoid using them inside of procedures. These steps do not seem like real steps - why is there "should" in the second step? It must be obvious what needs to be done. See "should" in Google style guide: https://developers.google.com/style/word-list?hl=en#should

For example:
`2. This file should contain the expected result of applying autofixes to the test code file.`
I need directive action inside of procedure. Do we actually want to say, that a user needs to: 
`2. In the test file, include a code that is detected and fixed by the autofix rule.`

Just an illustrative example (sorry if it sounds too simple): These manuals need to be the same as cooking recipes. "When the omelet is still liquid, put some mushrooms on the top of them." Not: "The liquid omelet should contain mushrooms." Ok, so it does not have include mushrooms, only should include them? Should I buy the eggs which already have mushrooms inside (contain?)? Do such eggs exist? Or I need to add them? What should I do? What a difficult recipe. :D

Similarly:
`3. Any future runs of the command `semgrep --test` will apply the autofix to the original testcode (`path/to/rule.py`) and verify whether this matches the expected outcome defined in the autofix test file (`path/to/rule.fixed.py`).
   If there is a mismatch, the line diffs will be printed.`
This is actually not a step with an action, but a comment after the user created an autofix rule, right? Removing it from the numbered list, maybe we can change it to:
`When you use `semgrep --test` again, Semgrep applies the autofix rule to the original test code (`path/to/rule.py`) and verifies whether this matches the expected outcome defined in the autofix test file (`path/to/rule.fixed.py`). If there is a mismatch, the line diffs are printed.`

I am still not sure what do we mean by the last comment - I would probably need to test it. What exactly we mean by a mismatch here is not clear to me. I am curious to test this out.

Reply:
Also: Will is more or less fine if there is an action that does not immediately lead to a result, for example "When you send an email, you will receive the answer between 8 or 24 hours." But a user can use `semgrep --test` immediately, so it is better to avoid the use of will here also.

Reply:
@s-santillan I see! One hour after I posted this comment according to GitHub. Thank you. :)